### PR TITLE
Operational APIs examples

### PR DESCRIPTION
**Pull request by htrant**
Demo of using the Operational APIs (consisting of the Equipment Status and Service Info APIs)

Equipment Status API 
- Fetch basic information of an equipment
- Fetch maintenance status of the equipment

Service Info API
- Fetch list of service orders for an equipment
- Fetch detailed information of a service order

### CODE REVIEW

File: src/common/koneapi.ts [hunk 1]
```diff
@@ -11,12 +11,15 @@ import {
...
+  ServiceOrder,
+  EquipmentStatus,
+  EquipmentInfo,
 } from './types'
 
 /**
  * Variables that contain the main endpoints used in this demo project.
  */
-const API_HOSTNAME = process.env.API_HOSTNAME || 'dev.kone.com'
+const API_HOSTNAME = process.env.API_HOSTNAME || 'dev.koneapi.com'
```
Comment (EitZei) [inline src/common/koneapi.ts]: Fix hostname

File: src/common/types.ts [hunk 2]
```diff
@@ -333,3 +333,38 @@ export class WebSocketSession extends EventEmitter {
    */
   sessionId?: string
 }
+
+/**
+ * Basic information of equipment
+ */
+export interface EquipmentInfo {
```
Comment (EitZei) [inline src/common/types.ts:340]: Check the optional fields. Now these types indicate that all the fields would be always present.

File: src/common/koneapi.ts [hunk 3]
```diff
@@ -447,3 +451,79 @@ export async function waitForResponse(
     webSocketConnection.prependListener('message', onMessage)
   })
 }
+
+/**
+ * Function to fetch basic information of an equipment
+ * @param accessToken valid access token
+ * @param equipmentIdWithPrefix equipment identifier with ken prefix
+ */
+export async function fetchEquipmentBasicInformation(accessToken: AccessToken, equipmentIdWithPrefix: string): Promise<EquipmentInfo> {
```
Comment (EitZei) [inline src/common/koneapi.ts]: Should these be split into a different file?

File: src/examples/operational-apis-demo.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import _ from 'lodash'
+import { fetchAccessToken, fetchEquipmentBasicInformation, fetchEquipmentStatus, fetchServiceOrdersList, fetchSingleServiceOrder } from '../common/koneapi'
+
+/**
+ * Update these three variables with your own credentials and equipment identifier or set them up as environment variables.
+ */
+const CLIENT_ID: string = process.env.CLIENT_ID || 'YOUR_CLIENT_ID' // eg. 'dcf48ab0-a902-4b52-8c53-1a9aede716e5'
+const CLIENT_SECRET: string = process.env.CLIENT_SECRET || 'YOUR_CLIENT_SECRET' // eg. '31d1329f8344fc12b1a960c8b8e0fc6a22ea7c35774c807a4fcabec4ffc8ae5b'
+const EQUIPMENT_ID_WITH_PREFIX: string = process.env.EQUIPMENT_ID_WITH_PREFIX || 'YOUR_EQUIPMENT_ID_WITH_KEN_PREFIX' // eg. sanbox equipment ken:9999999999
```
Comment (EitZei) [inline src/examples/operational-apis-demo.ts]: Should the be optional and use the inventory endpoint instead as default?

File: src/examples/operational-apis-demo.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  // Fetch basic information of equipment
+  console.log(`Fetch basic information of equipment ${EQUIPMENT_ID_WITH_PREFIX}`)
+  const equipmentInfo = await fetchEquipmentBasicInformation(accessToken, EQUIPMENT_ID_WITH_PREFIX)
+  console.log(equipmentInfo)
+
+  // Fetch maintenance status of equipment
+  console.log(`Fetch maintenance status of equipment ${EQUIPMENT_ID_WITH_PREFIX}`)
+  const equipmentStatus = await fetchEquipmentStatus(accessToken, EQUIPMENT_ID_WITH_PREFIX)
+  console.log(equipmentStatus)
```
Comment (EitZei) [inline src/examples/operational-apis-demo.ts:27]: Do the deep structures get rendered correctly or should this do `JSON.stringify(o, null, 4)`?

File: src/common/koneapi.ts (hunk 3 above)
Comment (mikko-kohtala) [reply src/common/koneapi.ts]: IMHO yes

Comment (mikko-kohtala) [reply src/common/koneapi.ts]: Also, these demos are lacking on the error-catching side quite a bit. Maybe try-catch around the axios requests?

**Review by mikko-kohtala — CHANGES_REQUESTED** [review]
Added a couple of smaller comments. Also, update the README.md and add PR description.

Once you have done these and Antti's proposed changes, please re-request review

File: src/examples/operational-apis-demo.ts (hunk 5 above)
Comment (htrant) [reply src/examples/operational-apis-demo.ts:27]: yes, it is rendered correctly. The object structure is not deep enough to cause issue :)

**Review by mikko-kohtala — CHANGES_REQUESTED** [review]
Thoughts / opinions / suggestions:

* We should try to keep the example file small and clear as possible. Meaning, that we could hide error catching in the common side, instead of the demo file. 
* demo file (start function) does all of the relevant console logging, which guides users through the application and gives information. Even when not running the application, the user would get explained what happens.
* src/common/operationalapi.ts  => operational-apis-supporting-functions.ts
* checkRequiredVariables is already in the common, and no need to catch that IMHO.

File: src/common/operational-api-supporting-functions.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
+import axios, { AxiosRequestConfig } from "axios"
+import { AccessToken, EquipmentInfo, EquipmentStatus, ServiceOrder } from "./types"
+
+const API_HOSTNAME = process.env.API_HOSTNAME || 'dev.kone.com'
+const API_EQUIPMENT_ENDPOINT = `https://${API_HOSTNAME}/api/v1/equipment`
+
+/**
+ * Function to fetch basic information of an equipment
+ * @param accessToken valid access token
+ * @param equipmentIdWithPrefix equipment identifier with ken prefix
```
Comment (gmokki) [inline src/common/operational-api-supporting-functions.ts]: Should we just call it `equipmentId` to match what the apis return?

File: src/common/operational-api-supporting-functions.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ */
+ export async function fetchEquipmentBasicInformation(accessToken: AccessToken, equipmentIdWithPrefix: string): Promise<EquipmentInfo> {
```
Comment (gmokki) [inline src/common/operational-api-supporting-functions.ts]: All these functions are 99% identical. You could combine the with a private helper function that accepts the acessToken, url and error message as parameters.

File: src/examples/operational-apis-demo.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
+import _ from 'lodash'
```
Comment (gmokki) [inline src/examples/operational-apis-demo.ts]: the only thing you use is the `isEmpty`, should we just import that one specifically?

**Review by EitZei — APPROVED** [review]
Looking good. Tested out the example with sandbox credentials and it worked perfectly.

File: src/examples/operational-apis-demo.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Update these two variables with your own credentials and equipment identifier or set them up as environment variables.
+ */
+const CLIENT_ID: string = process.env.CLIENT_ID || 'YOUR_CLIENT_ID' // eg. 'dcf48ab0-a902-4b52-8c53-1a9aede716e5'
+const CLIENT_SECRET: string = process.env.CLIENT_SECRET || 'YOUR_CLIENT_SECRET' // eg. '31d1329f8344fc12b1a960c8b8e0fc6a22ea7c35774c807a4fcabec4ffc8ae5b'
+
+/**
+ * Check if the needed variables have been defined
+ */
+const checkRequiredVariables = () => {
```
Comment (gmokki) [inline src/examples/operational-apis-demo.ts]: This is third identical definition of the checkRequiredVariables. Please use the `validateClientIdAndClientSecret` from `common/koneapi.ts` instead. And you could make the other examples use that too?

**Review by gmokki — APPROVED** [review]

**Review by mikko-kohtala — APPROVED** [review]

**Pull request merged**